## I.  Gathering Data


In [1]:
import numpy as np
import pandas as pd

In [2]:
#1. enhanced file
df = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [4]:
df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [6]:
#2. Data from API
df_json = pd.read_json('tweet_json2.txt', orient = 'records', lines = True)

In [7]:
df_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2337 entries, 0 to 2336
Data columns (total 32 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2337 non-null datetime64[ns]
display_text_range               2337 non-null object
entities                         2337 non-null object
extended_entities                2063 non-null object
favorite_count                   2337 non-null int64
favorited                        2337 non-null bool
full_text                        2337 non-null object
geo                              0 non-null float64
id                               2337 non-null int64
id_str                           2337 non-null int64
in_reply_to_screen_name          77 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        77 non-null float64
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          77 n

In [8]:
df_cleanjson = df_json[['id', 'favorite_count', 'retweet_count']]

In [9]:
df_cleanjson.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2337 entries, 0 to 2336
Data columns (total 3 columns):
id                2337 non-null int64
favorite_count    2337 non-null int64
retweet_count     2337 non-null int64
dtypes: int64(3)
memory usage: 54.9 KB


In [10]:
# 3. Image data
df_image = pd.read_csv('image_predictions.tsv',sep = '\t')

In [11]:
df_image.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [12]:
df_image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


## II. Accessing Data

In [13]:
# 1.Enhanced data
df

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [15]:
sum(df.expanded_urls.duplicated())

137

In [16]:
sum(df.source.duplicated())

2352

In [17]:
df.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [18]:
df.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [19]:
df_overnumerator = df.query('rating_numerator >= 15') 
df_overnumerator

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
55    881633300179243008           8.816070e+17         4.738443e+07   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
285   838916489579200512                    NaN                  NaN   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
340   832215909146226688                    NaN                  NaN   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
695   786709082849828864                    NaN                  NaN   
763   778027034220126208                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
979   749981277374128128                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1712  680494726643068929                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2074  670842764863651840                    NaN                  NaN   

                      timestamp  \
55    2017-07-02 21:58:53 +0000   
188   2017-04-22 19:15:32 +0000   
189   2017-04-22 19:05:32 +0000   
285   2017-03-07 00:57:32 +0000   
290   2017-03-04 22:12:52 +0000   
291   2017-03-04 17:56:49 +0000   
313   2017-02-24 21:54:03 +0000   
340   2017-02-16 13:11:49 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
695   2016-10-13 23:23:56 +0000   
763   2016-09-20 00:24:34 +0000   
902   2016-07-28 01:00:57 +0000   
979   2016-07-04 15:00:45 +0000   
1120  2016-05-13 16:15:54 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1663  2016-01-01 06:22:03 +0000   
1712  2015-12-25 21:06:00 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2074  2015-11-29 05:52:33 +0000   

                                                 source  \
55    <a href="http://twitter.com/download/iphone" r...   
188   <a href="http://twitter.com/download/iphone" r...   
189   <a href="http://twitter.com/download/iphone" r...   
285   <a href="http://twitter.com/download/iphone" r...   
290   <a href="http://twitter.com/download/iphone" r...   
291   <a href="http://twitter.com/download/iphone" r...   
313   <a href="http://twitter.com/download/iphone" r...   
340   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
979   <a href="https://about.twitter.com/products/tw...   
1120  <a href="http://twit

In [20]:
len(df_overnumerator)

28

In [21]:
df_overnumerator.text

55      @roushfenway These are good dogs but 17/10 is ...
188     @dhmontgomery We also gave snoop dogg a 420/10...
189     @s8n You tried very hard to portray this good ...
285     RT @KibaDva: I collected all the good dogs!! 1...
290                                    @markhoppus 182/10
291     @bragg6of8 @Andy_Pace_ we are still looking fo...
313     @jonnysun @Lin_Manuel ok jomny I know you're e...
340     RT @dog_rates: This is Logan, the Chow who liv...
433     The floofs have been released I repeat the flo...
516     Meet Sam. She smiles 24/7 &amp; secretly aspir...
695     This is Logan, the Chow who lived. He solemnly...
763     This is Sophie. She's a Jubilant Bush Pupper. ...
902     Why does this never happen at my front door......
979     This is Atticus. He's quite simply America af....
1120    Say hello to this unbelievably well behaved sq...
1202    This is Bluebert. He just saw that both #Final...
1228    Happy Saturday here's 9 puppers on a bench. 99...
1254    Here's

In [22]:
sum(df.expanded_urls.duplicated())

137

In [23]:
df.expanded_urls.isnull().sum()

59

In [24]:
df.name.value_counts()

None         745
a             55
Charlie       12
Lucy          11
Cooper        11
Oliver        11
Penny         10
Tucker        10
Lola          10
Bo             9
Winston        9
the            8
Sadie          8
Toby           7
an             7
Buddy          7
Daisy          7
Bailey         7
Jax            6
Stanley        6
Bella          6
Leo            6
Koda           6
Oscar          6
Milo           6
Jack           6
Scout          6
Dave           6
Rusty          6
Bentley        5
            ... 
Rover          1
Bradley        1
Apollo         1
Marvin         1
Tiger          1
Spencer        1
Grizzie        1
Fletcher       1
Acro           1
Spark          1
DayZ           1
Duddles        1
Goose          1
Grizz          1
Socks          1
Bruno          1
Herb           1
Fiji           1
Eugene         1
Baron          1
this           1
Jebberson      1
Liam           1
Philippe       1
Boots          1
such           1
Kallie         1
Smiley        

 # Quality
 
 - Timestamp datatype should change from object into date
 - Correct the right denominator 
 - Correct the right numerator  เกิดจากการใช้โค้ดดีเทคตัวเลขในคอมเม้นท์ ซึ่งบางคอมเม้นมันอวยเว่อร์เกิน (มักจะเป็นอันรีทวีต) หรือบางอันเช่น smile 24/7 ก็แปลงค่าเป็นคะแนน
 - Correct name : 'None' and 'a'.
 - เอาช่องที่เป็น retweet ออก
 - expanded_urls duplicated เกิดจาก RTตรงช่อง text น่าจะแก้ได้พร้อมกับการเอาช่องretweetออก
 - ช่อง 340	832215909146226688 จริงๆคะแนนคือ 9.75 -- หรือควรเปลี่ยน datatypeเป็น float?
 - แก้ in_reply_to_status_id ให้เป็นตัวเลขปกติ
 
 # Tidiness
 - Merge all dog states into one column

In [25]:
# 2. Json Data
df_cleanjson.head()

id  favorite_count  retweet_count
0  892420643555336193           37733           8222
1  892177421306343426           32416           6079
2  891815181378084864           24412           4023
3  891689557279858688           41063           8369
4  891327558926688256           39252           9082

In [26]:
df_cleanjson.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2337 entries, 0 to 2336
Data columns (total 3 columns):
id                2337 non-null int64
favorite_count    2337 non-null int64
retweet_count     2337 non-null int64
dtypes: int64(3)
memory usage: 54.9 KB


# Quality
    - Change the column name id into tweet_id

# Tidiness
    - 

In [27]:
#  3. Image
df_image.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [28]:
df_image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [29]:
sum(df_image.jpg_url.duplicated())

66

# Quality
   - Delete the duplicate jpg_url

# Tidiness
    - Merge table json with table image 

## Cleansing Data

> # 1. Enhanced table

 ># Quality
 
 - Timestamp datatype should change from object into date
 - Correct the right denominator 
 - Correct the right numerator  เกิดจากการใช้โค้ดดีเทคตัวเลขในคอมเม้นท์ ซึ่งบางคอมเม้นมันอวยเว่อร์เกิน (มักจะเป็นอันรีทวีต) หรือบางอันเช่น smile 24/7 ก็แปลงค่าเป็นคะแนน
 - Correct name : 'None' and 'a'. What can be done in this case is we can use string patterns like This is *, Meet *, named * So we can identify these 3-4 patterns and then use regex to get the words at * in code. This will clean many of the names
 - เอาช่องที่เป็น retweet ออก
 - expanded_urls duplicated เกิดจาก RTตรงช่อง text น่าจะแก้ได้พร้อมกับการเอาช่องretweetออก
 - ช่อง 340	832215909146226688 จริงๆคะแนนคือ 9.75 -- หรือควรเปลี่ยน datatypeเป็น float?
 - แก้ in_reply_to_status_id ให้เป็นตัวเลขปกติ
 - 835152434251116546 เป็นโพสต์ที่ประจานคนก๊อปโพสต์ของเค้าไปลง เค้าเลยให้คะแนน 0/10 
 - 670842764863651840 เป็นรูปคน (น่าจะสนูปด็อก = =)
 - 716439118184652801	คะแนนคืออันที่2: 11/10 (ไม่ใช่ 50/50)
 - 838150277551247360 น่าจะเป็นการ reply คะแนนตอบเพื่อน
 
 # Tidiness
 - Merge all dog states into one column

In [30]:
df_enhanced_copy = df.copy()

> 1.1 เปลี่ยน datastamp เป็น datetime data type

In [31]:
df_enhanced_copy['timestamp'] =  pd.to_datetime(df_enhanced_copy['timestamp'])

In [32]:
df_enhanced_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null datetime64[ns]
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: datetime64[ns](1

In [33]:
df_enhanced_copy.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

            timestamp                                             source  \
0 2017-08-01 16:23:56  <a href="http://twitter.com/download/iphone" r...   
1 2017-08-01 00:17:27  <a href="http://twitter.com/download/iphone" r...   
2 2017-07-31 00:18:03  <a href="http://twitter.com/download/iphone" r...   
3 2017-07-30 15:58:51  <a href="http://twitter.com/download/iphone" r...   
4 2017-07-29 16:00:24  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

> 1.2 rating_numerator และ rating_denominator

In [34]:
df_enhanced_copy.rating_numerator = df_enhanced_copy.text.str.extract('((?:\d+\.)?\d+)\/(\d+)', expand=True).astype(float)

In [35]:
df_enhanced_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null datetime64[ns]
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null float64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: datetime64[ns]

In [36]:
df_enhanced_copy_overnumerator = df_enhanced_copy.query('rating_numerator >= 15') 
df_enhanced_copy_overnumerator

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
55    881633300179243008           8.816070e+17         4.738443e+07   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
285   838916489579200512                    NaN                  NaN   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
979   749981277374128128                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2074  670842764863651840                    NaN                  NaN   

               timestamp                                             source  \
55   2017-07-02 21:58:53  <a href="http://twitter.com/download/iphone" r...   
188  2017-04-22 19:15:32  <a href="http://twitter.com/download/iphone" r...   
189  2017-04-22 19:05:32  <a href="http://twitter.com/download/iphone" r...   
285  2017-03-07 00:57:32  <a href="http://twitter.com/download/iphone" r...   
290  2017-03-04 22:12:52  <a href="http://twitter.com/download/iphone" r...   
291  2017-03-04 17:56:49  <a href="http://twitter.com/download/iphone" r...   
313  2017-02-24 21:54:03  <a href="http://twitter.com/download/iphone" r...   
433  2017-01-15 17:52:40  <a href="http://twitter.com/download/iphone" r...   
516  2016-12-19 23:06:23  <a href="http://twitter.com/download/iphone" r...   
902  2016-07-28 01:00:57  <a href="http://twitter.com/download/iphone" r...   
979  2016-07-04 15:00:45  <a href="https://about.twitter.com/products/tw...   
1120 2016-05-13 16:15:54  <a href="http://twitter.com/download/iphone" r...   
1202 2016-04-03 01:36:11  <a href="http://twitter.com/download/iphone" r...   
1228 2016-03-27 01:29:02  <a href="http://twitter.com/download/iphone" r...   
1254 2016-03-18 02:46:49  <a href="http://twitter.com/download/iphone" r...   
1274 2016-03-14 02:04:08  <a href="http://twitter.com/download/iphone" r...   
1351 2016-02-28 21:25:30  <a href="http://twitter.com/download/iphone" r...   
1433 2016-02-10 16:51:59  <a href="http://twitter.com/download/iphone" r...   
1634 2016-01-05 04:11:44  <a href="http://twitter.com/download/iphone" r...   
1635 2016-01-05 04:00:18  <a href="http://twitter.com/download/iphone" r...   
1663 2016-01-01 06:22:03  <a href="http://twitter.com/download/iphone" r...   
1779 2015-12-18 05:06:23  <a href="http://twitter.com/download/iphone" r...   
1843 2015-12-13 01:41:41  <a href="http://twitter.com/download/iphone" r...   
2074 2015-11-29 05:52:33  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
55    @roushfenway These are good dogs but 17/10 is ...                  NaN   
188   @dhmontgomery We also gave snoop dogg a 420/10...                  NaN   

In [37]:
df_enhanced_copy.query('name == "Logan"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
340  832215909146226688                    NaN                  NaN   
695  786709082849828864                    NaN                  NaN   

              timestamp                                             source  \
340 2017-02-16 13:11:49  <a href="http://twitter.com/download/iphone" r...   
695 2016-10-13 23:23:56  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
340  RT @dog_rates: This is Logan, the Chow who liv...         7.867091e+17   
695  This is Logan, the Chow who lived. He solemnly...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
340              4.196984e+09  2016-10-13 23:23:56 +0000   
695                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
340  https://twitter.com/dog_rates/status/786709082...              9.75   
695  https://twitter.com/dog_rates/status/786709082...              9.75   

     rating_denominator   name doggo floofer pupper puppo  
340                  10  Logan  None    None   None  None  
695                  10  Logan  None    None   None  None

> เอาข้อมูลที่เป็น reply กับ retweet ออก

In [38]:
df_enhanced_copy.in_reply_to_status_id.notnull().sum()

78

In [39]:
df_enhanced_copy = df_enhanced_copy.query('in_reply_to_status_id == "NaN"')

In [40]:
df_enhanced_copy.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

            timestamp                                             source  \
0 2017-08-01 16:23:56  <a href="http://twitter.com/download/iphone" r...   
1 2017-08-01 00:17:27  <a href="http://twitter.com/download/iphone" r...   
2 2017-07-31 00:18:03  <a href="http://twitter.com/download/iphone" r...   
3 2017-07-30 15:58:51  <a href="http://twitter.com/download/iphone" r...   
4 2017-07-29 16:00:24  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...              13.0   
1  https://twitter.com/dog_rates/status/892177421...              13.0   
2  https://twitter.com/dog_rates/status/891815181...              12.0   
3  https://twitter.com/dog_rates/status/891689557...              13.0   
4  https://twitter.com/dog_rates/status/891327558...              12.0   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [41]:
df_enhanced_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2278 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2278 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2278 non-null datetime64[ns]
source                        2278 non-null object
text                          2278 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2274 non-null object
rating_numerator              2278 non-null float64
rating_denominator            2278 non-null int64
name                          2278 non-null object
doggo                         2278 non-null object
floofer                       2278 non-null object
pupper                        2278 non-null object
puppo                         2278 non-null object
dtypes: datetime64[ns](1

In [42]:
df_enhanced_copy.retweeted_status_id.notnull().sum()

181

In [43]:
df_enhanced_copy = df_enhanced_copy.query('retweeted_status_id == "NaN"')

In [44]:
df_enhanced_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2097 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2097 non-null datetime64[ns]
source                        2097 non-null object
text                          2097 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2094 non-null object
rating_numerator              2097 non-null float64
rating_denominator            2097 non-null int64
name                          2097 non-null object
doggo                         2097 non-null object
floofer                       2097 non-null object
pupper                        2097 non-null object
puppo                         2097 non-null object
dtypes: datetime64[ns](1), flo

In [45]:
df_enhanced_copy_overnumerator = df_enhanced_copy.query('rating_numerator >= 15') 
df_enhanced_copy_overnumerator

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
979   749981277374128128                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1635  684222868335505415                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2074  670842764863651840                    NaN                  NaN   

               timestamp                                             source  \
433  2017-01-15 17:52:40  <a href="http://twitter.com/download/iphone" r...   
516  2016-12-19 23:06:23  <a href="http://twitter.com/download/iphone" r...   
902  2016-07-28 01:00:57  <a href="http://twitter.com/download/iphone" r...   
979  2016-07-04 15:00:45  <a href="https://about.twitter.com/products/tw...   
1120 2016-05-13 16:15:54  <a href="http://twitter.com/download/iphone" r...   
1202 2016-04-03 01:36:11  <a href="http://twitter.com/download/iphone" r...   
1228 2016-03-27 01:29:02  <a href="http://twitter.com/download/iphone" r...   
1254 2016-03-18 02:46:49  <a href="http://twitter.com/download/iphone" r...   
1274 2016-03-14 02:04:08  <a href="http://twitter.com/download/iphone" r...   
1351 2016-02-28 21:25:30  <a href="http://twitter.com/download/iphone" r...   
1433 2016-02-10 16:51:59  <a href="http://twitter.com/download/iphone" r...   
1635 2016-01-05 04:00:18  <a href="http://twitter.com/download/iphone" r...   
1779 2015-12-18 05:06:23  <a href="http://twitter.com/download/iphone" r...   
1843 2015-12-13 01:41:41  <a href="http://twitter.com/download/iphone" r...   
2074 2015-11-29 05:52:33  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
433   The floofs have been released I repeat the flo...                  NaN   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   
902   Why does this never happen at my front door......                  NaN   
979   This is Atticus. He's quite simply America af....                  NaN   
1120  Say hello to this unbelievably well behaved sq...                  NaN   
1202  This is Bluebert. He just saw that both #Final...                  NaN   
1228  Happy Saturday here's 9 puppers on a bench. 99...                  NaN   
1254  Here's a brigade of puppers. All look very pre...                  NaN   
1274  From left to right:\nCletus, Jerome, Alejandro...                  NaN   
1351  Here is a whole flock of puppers.  60/50 I'll ...                  NaN   
1433  Happy Wednesday here's a bucket of pups. 44/40...                  NaN   
1635  Someone help the girl is being mugged. Several...                  NaN   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...                  NaN   
1843  Here we have an entire platoon of puppers. Tot...                  NaN   
2074  After so many requests... here you go.\n\nGood...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
433                        NaN                        NaN   
516                        NaN                        NaN   
902                        NaN                        NaN   
979                        NaN                        NaN   
1120        

In [46]:
len(df_enhanced_copy_overnumerator)

15

In [47]:
df_enhanced_copy.query('name == "Logan"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
695  786709082849828864                    NaN                  NaN   

              timestamp                                             source  \
695 2016-10-13 23:23:56  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
695  This is Logan, the Chow who lived. He solemnly...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
695                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
695  https://twitter.com/dog_rates/status/786709082...              9.75   

     rating_denominator   name doggo floofer pupper puppo  
695                  10  Logan  None    None   None  None

> duplicated expanded url

In [48]:
sum(df_enhanced_copy.expanded_urls.duplicated())

3

In [49]:
dup_url = df_enhanced_copy[df_enhanced_copy.duplicated(['expanded_urls'], keep = False)]

In [50]:
dup_url.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5 entries, 375 to 2212
Data columns (total 17 columns):
tweet_id                      5 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     5 non-null datetime64[ns]
source                        5 non-null object
text                          5 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2 non-null object
rating_numerator              5 non-null float64
rating_denominator            5 non-null int64
name                          5 non-null object
doggo                         5 non-null object
floofer                       5 non-null object
pupper                        5 non-null object
puppo                         5 non-null object
dtypes: datetime64[ns](1), float64(5), int64(2), object(9)
memory u

In [51]:
dup_url = dup_url.query('expanded_urls != "NaN"')

In [52]:
dup_url

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
375   828361771580813312                    NaN                  NaN   
657   791774931465953280                    NaN                  NaN   
707   785515384317313025                    NaN                  NaN   
1445  696518437233913856                    NaN                  NaN   
2212  668587383441514497                    NaN                  NaN   

               timestamp                                             source  \
375  2017-02-05 21:56:51  <a href="http://twitter.com" rel="nofollow">Tw...   
657  2016-10-27 22:53:48  <a href="http://vine.co" rel="nofollow">Vine -...   
707  2016-10-10 16:20:36  <a href="http://twitter.com/download/iphone" r...   
1445 2016-02-08 02:18:30  <a href="http://twitter.com/download/iphone" r...   
2212 2015-11-23 00:30:28  <a href="http://vine.co" rel="nofollow">Vine -...   

                                                   text  retweeted_status_id  \
375   Beebop and Doobert should start a band 12/10 w...                  NaN   
657   Vine will be deeply missed. This was by far my...                  NaN   
707      Today, 10/10, should be National Dog Rates Day                  NaN   
1445    Oh my god 10/10 for every little hot dog pupper                  NaN   
2212  Never forget this vine. You will not stop watc...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
375                        NaN                        NaN   
657                        NaN                        NaN   
707                        NaN                        NaN   
1445                       NaN                        NaN   
2212                       NaN                        NaN   

                      expanded_urls  rating_numerator  rating_denominator  \
375                             NaN              12.0                  10   
657   https://vine.co/v/ea0OwvPTx9l              14.0                  10   
707                             NaN              10.0                  10   
1445                            NaN              10.0                  10   
2212  https://vine.co/v/ea0OwvPTx9l              13.0                  10   

      name doggo floofer  pupper puppo  
375   None  None    None    None  None  
657   None  None    None    None  None  
707   None  None    None    None  None  
1445  None  None    None  pupper  None  
2212   the  None    None    None  None

> Tweet id 791774931465953280 and 668587383441514497 are the same dog. They are rated in different scores and different years. Thus, I'll keep only the oldest one.

In [53]:
df_enhanced_copy.drop(df_enhanced_copy.tweet_id == 668587383441514497, inplace = True)

KeyError: '[False False False ... False False False] not found in axis'

In [54]:
df_enhanced_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2097 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2097 non-null datetime64[ns]
source                        2097 non-null object
text                          2097 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2094 non-null object
rating_numerator              2097 non-null float64
rating_denominator            2097 non-null int64
name                          2097 non-null object
doggo                         2097 non-null object
floofer                       2097 non-null object
pupper                        2097 non-null object
puppo                         2097 non-null object
dtypes: datetime64[ns](1), flo

> Correct name : 'None' and 'a'. What can be done in this case is we can use string patterns like This is *, Meet *, named * So we can identify these 3-4 patterns and then use regex to get the words at * in code. This will clean many of the names

In [55]:
df_enhanced_copy.name.value_counts()

None         603
a             55
Lucy          11
Charlie       11
Oliver        10
Cooper        10
Penny          9
Tucker         9
Sadie          8
Winston        8
the            8
Lola           8
Toby           7
Daisy          7
Bella          6
an             6
Koda           6
Oscar          6
Bailey         6
Jax            6
Stanley        6
Bo             6
Louis          5
Scout          5
Milo           5
Dave           5
Rusty          5
Buddy          5
Bentley        5
Leo            5
            ... 
Stephan        1
Gin            1
Heinrich       1
Rover          1
Jeremy         1
Smiley         1
Kallie         1
such           1
Dudley         1
Gilbert        1
Pubert         1
Shelby         1
Dylan          1
Dot            1
Spark          1
DayZ           1
Duddles        1
Goose          1
Grizz          1
Socks          1
Bruno          1
Herb           1
Fiji           1
Acro           1
Eugene         1
this           1
Jebberson      1
Liam          

> rating_denominator

In [56]:
df_enhanced_copy.rating_denominator.value_counts()

10     2080
50        3
11        2
80        2
7         1
170       1
150       1
120       1
110       1
90        1
70        1
40        1
20        1
2         1
Name: rating_denominator, dtype: int64

In [57]:
df_enhanced_copy_baddenominator = df_enhanced_copy.query('rating_denominator != 10') 
df_enhanced_copy_baddenominator

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1165  722974582966214656                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1635  684222868335505415                    NaN                  NaN   
1662  682962037429899265                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

               timestamp                                             source  \
433  2017-01-15 17:52:40  <a href="http://twitter.com/download/iphone" r...   
516  2016-12-19 23:06:23  <a href="http://twitter.com/download/iphone" r...   
902  2016-07-28 01:00:57  <a href="http://twitter.com/download/iphone" r...   
1068 2016-06-08 02:41:38  <a href="http://twitter.com/download/iphone" r...   
1120 2016-05-13 16:15:54  <a href="http://twitter.com/download/iphone" r...   
1165 2016-04-21 02:25:47  <a href="http://twitter.com/download/iphone" r...   
1202 2016-04-03 01:36:11  <a href="http://twitter.com/download/iphone" r...   
1228 2016-03-27 01:29:02  <a href="http://twitter.com/download/iphone" r...   
1254 2016-03-18 02:46:49  <a href="http://twitter.com/download/iphone" r...   
1274 2016-03-14 02:04:08  <a href="http://twitter.com/download/iphone" r...   
1351 2016-02-28 21:25:30  <a href="http://twitter.com/download/iphone" r...   
1433 2016-02-10 16:51:59  <a href="http://twitter.com/download/iphone" r...   
1635 2016-01-05 04:00:18  <a href="http://twitter.com/download/iphone" r...   
1662 2016-01-01 16:30:13  <a href="http://twitter.com/download/iphone" r...   
1779 2015-12-18 05:06:23  <a href="http://twitter.com/download/iphone" r...   
1843 2015-12-13 01:41:41  <a href="http://twitter.com/download/iphone" r...   
2335 2015-11-16 16:11:11  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
433   The floofs have been released I repeat the flo...                  NaN   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   
902   Why does this never happen at my front door......                  NaN   
1068  After so many requests, this is Bretagne. She ...                  NaN   
1120  Say hello to this unbelievably well behaved sq...                  NaN   
1165  Happy 4/20 from the squad! 13/10 for all https...                  NaN   
1202  This is Bluebert. He just saw that both #Final...                  NaN   
1228  Happy Saturday here's 9 puppers on a bench. 99...                  NaN   
1254  Here's a brigade of puppers. All look very pre...                  NaN   
1274  From left to right:\nCletus, Jerome, Alejandro...                  NaN   
1351  Here is a whole flock of puppers.  60/50 I'll ...                  NaN   
1433  Happy Wednesday here's a bucket of pups. 44/40...                  NaN   
1635  Someone help the girl is being mugged. Several...                  NaN   
1662  This is Darrel. He just robbed a 7/11 and is i...                  NaN   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...                  NaN   
1843  Here we ha

In [59]:
def find_nth(haystack, needle, n):
    start = haystack.find(needle)
    while start >= 0 and n > 1:
        start = haystack.find(needle, start+len(needle))
        n -= 1
    return start

In [60]:
find_nth(text, '/', 2)

AttributeError: 'Series' object has no attribute 'find'

In [84]:
df_enhanced_copy

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
31    886258384151887873                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

# 2. Json Data

# Quality
    - Change the column name id into tweet_id

# Tidiness
    - 

> - Change the column name id into tweet_id

In [61]:
df_cleanjson_copy =  df_cleanjson.copy()

In [62]:
df_cleanjson_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2337 entries, 0 to 2336
Data columns (total 3 columns):
id                2337 non-null int64
favorite_count    2337 non-null int64
retweet_count     2337 non-null int64
dtypes: int64(3)
memory usage: 54.9 KB


In [63]:
df_cleanjson_copy.rename(index = str, columns = {'id': 'tweet_id'}, inplace =  True)

In [64]:
df_cleanjson_copy.head()

tweet_id  favorite_count  retweet_count
0  892420643555336193           37733           8222
1  892177421306343426           32416           6079
2  891815181378084864           24412           4023
3  891689557279858688           41063           8369
4  891327558926688256           39252           9082

In [65]:
df_cleanjson_copy.describe()

tweet_id  favorite_count  retweet_count
count  2.337000e+03     2337.000000    2337.000000
mean   7.421828e+17     7892.863928    2901.230638
std    6.830233e+16    12231.117156    4891.523367
min    6.660209e+17        0.000000       0.000000
25%    6.783411e+17     1364.000000     583.000000
50%    7.186133e+17     3433.000000    1353.000000
75%    7.986946e+17     9664.000000    3380.000000
max    8.924206e+17   162867.000000   82881.000000

# 3.Image

> Quality
 - Delete the duplicate jpg_url
 
 >Tidiness
- Merge table json with table image 

In [66]:
df_image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [67]:
df_image_copy = df_image.copy()

In [68]:
df_image_copy["is_duplicate"]= df_image_copy.jpg_url.duplicated()

In [77]:
df_image_copy.query('is_duplicate == False',inplace = True)

In [82]:
df_image_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2009 entries, 0 to 2074
Data columns (total 13 columns):
tweet_id        2009 non-null int64
jpg_url         2009 non-null object
img_num         2009 non-null int64
p1              2009 non-null object
p1_conf         2009 non-null float64
p1_dog          2009 non-null bool
p2              2009 non-null object
p2_conf         2009 non-null float64
p2_dog          2009 non-null bool
p3              2009 non-null object
p3_conf         2009 non-null float64
p3_dog          2009 non-null bool
is_duplicate    2009 non-null bool
dtypes: bool(4), float64(3), int64(2), object(4)
memory usage: 164.8+ KB
